### PR TITLE
feat: support keyword text match

### LINKED ISSUE #35922: [Feature]: keyword text match
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe.

Not related to a problem.

### Describe the solution you'd like.

I want to support keyword text match in Milvus. With this feature enabled in the schema, the Milvus user can easily do the keyword matching. For example:

```python
import pymilvus

from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection,
)

DIMENSION = 8
COLLECTION_NAME = "books"
connections.connect("default", host="localhost", port="19530")

fields = [
    FieldSchema(name='id', dtype=DataType.INT64, is_primary=True),
    FieldSchema(name='title', dtype=DataType.VARCHAR, max_length=200, enable_match=True),
    FieldSchema(name='release_year', dtype=DataType.INT64),
    FieldSchema(name='embeddings', dtype=DataType.FLOAT_VECTOR, dim=DIMENSION),
]
schema = CollectionSchema(fields=fields, enable_dynamic_field=True)
collection = Collection(name=COLLECTION_NAME, schema=schema)

data_rows = [
    {
        "id": 1,
        "title": "Lord of the Flies",
        "release_year": 1954,
        "embeddings": [0.64, 0.44, 0.13, 0.47, 0.74, 0.03, 0.32, 0.6],
    },
    {
        "id": 2,
        "title": "The Great Gatsby",
        "release_year": 1925,
        "embeddings": [0.9, 0.45, 0.18, 0.43, 0.4, 0.4, 0.7, 0.24],
    },
    {
        "id": 3,
        "title": "The Catcher in the Rye",
        "release_year": 1951,
        "embeddings": [0.43, 0.57, 0.43, 0.88, 0.84, 0.69, 0.27, 0.98],
    },
    {
        "id": 4,
        "title": "Flipped",
        "release_year": 2010,
        "embeddings": [0.84, 0.69, 0.27, 0.43, 0.57, 0.43, 0.88, 0.98],
    },
]

collection.insert(data_rows)
collection.create_index(
    "embeddings", {"index_type": "FLAT", "metric_type": "L2"})

collection.load()
collection.flush()

r = collection.query(expr="TextMatch(title, 'The')", output_fields=["id", "title"])
print(r)

r = collection.query(expr="not TextMatch(title, 'The')", output_fields=["id", "title"])
print(r)

r = collection.query(expr="TextMatch(title, 'The Catcher')", output_fields=["id", "title"])
print(r)

r = collection.query(expr="TextMatch(title, 'The') and TextMatch(title, 'Catcher')", output_fields=["id", "title"])
print(r)

collection.drop()

```

### Describe an alternate solution.

_No response_

### Anything else? (Additional Context)

_No response_

### PR DESCRIPTION
**Pull request by longjiquan**
fix: #35922

### CODE REVIEW

File: internal/core/src/index/TextMatchIndex.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    auto cnt = wrapper_->count();
+    TargetBitmap bitset(cnt);
+    if (bitset.empty()) {
+        return bitset;
+    }
+    auto hits = wrapper_->match_query(query);
+    apply_hits(bitset, hits, true);
+    return bitset;
```
Comment (sunby) [inline internal/core/src/index/TextMatchIndex.cpp:166]: will all hits be less than cnt? Is there a conner case that inverted index has 10 rows but we actually have 11 rows so hits may return an index 10.

File: internal/core/src/index/InvertedIndexUtil.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#pragma once
+
+namespace milvus::index {
+inline void
+apply_hits(milvus::TargetBitmap& bitset,
+           const milvus::index::RustArrayWrapper& w,
+           bool v) {
+    for (size_t j = 0; j < w.array_.len; j++) {
```
Comment (sunby) [inline internal/core/src/index/InvertedIndexUtil.h:19]: maybe better to use size_t i

File: internal/core/src/exec/expression/Expr.h [hunk 3]
```diff
@@ -323,6 +323,34 @@ class SegmentExpr : public Expr {
...
+        TargetBitmap result;
+
+        if (cached_match_res_ == nullptr) {
+            auto index = segment_->GetTextIndex(field_id_);
+            auto res = std::move(func(index, values...));
+            cached_match_res_ = std::make_shared<TargetBitmap>(std::move(res));
+            if (cached_match_res_->size() < active_count_) {
+                // some entities are not visible in inverted index.
+                TargetBitmap tail(active_count_ - cached_match_res_->size());
+                cached_match_res_->append(tail);
```
Comment (sunby) [inline internal/core/src/exec/expression/Expr.h:338]: is it right to just append 0s to result?

File: internal/core/src/exec/expression/UnaryExpr.cpp [hunk 4]
```diff
@@ -601,6 +601,10 @@ PhyUnaryRangeFilterExpr::ExecRangeVisitorImplJson() {
 template <typename T>
 VectorPtr
 PhyUnaryRangeFilterExpr::ExecRangeVisitorImpl() {
+    if (expr_->op_type_ == proto::plan::OpType::TextMatch) {
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/src/exec/expression/UnaryExpr.cpp:604]: OK.

File: internal/core/src/index/TextMatchIndex.cpp [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+TextMatchIndex::CreateReader() {
+    wrapper_->create_reader();
+}
+
+TargetBitmap
+TextMatchIndex::MatchQuery(const std::string& query) {
+    if (shouldTriggerCommit()) {
+        Commit();
+        Reload();
+    }
```
[the remark this answers is not in the document]
Comment (longjiquan) [reply internal/core/src/index/TextMatchIndex.cpp:157]: A worker thread per segment will introduce a lot of overhead. Maybe we can use a thread pool or a special thread to do this.

File: internal/core/src/index/TextMatchIndex.cpp (hunk 1 above)
Comment (longjiquan) [reply internal/core/src/index/TextMatchIndex.cpp:166]: Yeah, some entities of growing segment may be not visible if they are not commited.

File: internal/core/src/index/InvertedIndexUtil.h (hunk 2 above)
Comment (longjiquan) [reply internal/core/src/index/InvertedIndexUtil.h:19]: OK.

File: internal/core/src/exec/expression/Expr.h (hunk 3 above)
Comment (longjiquan) [reply internal/core/src/exec/expression/Expr.h:338]: Yeah, some entities of growing segment may be not visible if they are not commited.

File: pkg/util/typeutil/field_schema.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+func (h *FieldSchemaHelper) EnableMatch() bool {
+	if !IsStringType(h.schema.GetDataType()) {
+		return false
+	}
+	s, err := h.typeParams.Get("enable_match")
+	if err != nil {
+		return false
+	}
+	enable, err := strconv.ParseBool(s)
```
Comment (czs007) [inline pkg/util/typeutil/field_schema.go:52]: so, invalid bool is considered as False?

Comment (longjiquan) [reply pkg/util/typeutil/field_schema.go:52]: yes

File: internal/core/src/index/TextMatchIndex.cpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+void
+TextMatchIndex::CreateReader() {
+    wrapper_->create_reader();
+}
+
+TargetBitmap
+TextMatchIndex::MatchQuery(const std::string& query) {
+    if (shouldTriggerCommit()) {
```
Comment (czs007) [inline internal/core/src/index/TextMatchIndex.cpp:154]: what happens if concurrent Commit and Reload is triggered?

Comment (longjiquan) [reply internal/core/src/index/TextMatchIndex.cpp:154]: try lock optimization can be applied.

File: internal/core/src/index/TextMatchIndex.cpp [hunk 8]
```diff
@@ -135,8 +151,10 @@ TextMatchIndex::shouldTriggerCommit() {
 
 void
 TextMatchIndex::Commit() {
-    wrapper_->commit();
-    last_commit_time_.store(stdclock::now());
+    if (mtx_.try_lock()) {
+        wrapper_->commit();
+        last_commit_time_.store(stdclock::now());
```
Comment (czs007) [inline internal/core/src/index/TextMatchIndex.cpp:156]: need call mtx_.unlock()

Comment (czs007) [reply internal/core/src/index/TextMatchIndex.cpp:156]: Use unique_lock is better.
[std::unique_lock](http://en.cppreference.com/w/cpp/thread/unique_lock)<[std::mutex](http://en.cppreference.com/w/cpp/thread/mutex)> lock(mtx_, [std::defer_lock](http://en.cppreference.com/w/cpp/thread/lock_tag));
if lock.try_lock() {
...
}

Comment (czs007) [reply internal/core/src/index/TextMatchIndex.cpp:156]: It seems that we can also use [std::recursive_mutex](http://en.cppreference.com/w/cpp/thread/recursive_mutex) to ensure safety

Comment (longjiquan) [reply internal/core/src/index/TextMatchIndex.cpp:157]: fixed

**Pull request merged**